kernel/kernel-api/src/main/java/io/delta/kernel/internal/util/SchemaIterable.java

```diff
@@ -0,0 +1,450 @@
+    private final Consumer<StructType> finalizedSchemaConsumer;
+    private SchemaZipper nextZipper;
+    private boolean finishedVisitingCurrent = false;
+
+    SchemaIterator(StructType schema, Consumer<StructType> finalizedSchemaConsumer) {
+      this.nextZipper = SchemaZipper.createZipper(schema);
+      this.finalizedSchemaConsumer = finalizedSchemaConsumer;
+      // Special case if struct is empty no force no iteration.
+      this.finishedVisitingCurrent = schema.fields().isEmpty();
+    }
+
+    @Override
+    public boolean hasNext() {
+      boolean nextAvailable =
+          nextZipper != null
+              && (!finishedVisitingCurrent // Implies there are children.
+                  // Without children there must be siblings or parents left to visit to have a next
+                  // value.
+                  || nextZipper.hasMoreSiblings()
+                  || nextZipper.hasParents());
+      if (!nextAvailable && nextZipper != null) {
+        finalizedSchemaConsumer.accept((StructType) nextZipper.extractDataTypeFromFields());
+        nextZipper = null;
+      }
+      return nextAvailable;
+    }
+
+    @Override
+    public MutableSchemaElement next() {
+      if (!hasNext()) {
+        throw new NoSuchElementException();
+      }
+      advanceNext();
+      return nextZipper;
+    }
+
+    private void advanceNext() {
+      while (nextZipper != null) {
+        if (!finishedVisitingCurrent) {
+          if (nextZipper.hasChildren()) {
+            // Try to go deeper if we haven't visited this node yet
+            while (nextZipper.hasChildren()) {
+              nextZipper = nextZipper.childrenZipper();
+            }
+          }
+          // At a leaf node, so by definition it is finished visiting.
+          finishedVisitingCurrent = true;
+          return;
+        }
+
+        // Try moving to sibling if there is no need to go down further.
+        if (nextZipper.hasMoreSiblings()) {
+          nextZipper = nextZipper.moveToSibling();
+          if (nextZipper.hasChildren()) {
+            // Force visiting children first.
+            finishedVisitingCurrent = false;
+            continue;
+          }
+          finishedVisitingCurrent = true;
+          return;
+        }
+
+        // Last remaining direction with no children and no
+        // siblings is to pop back up and note that visiting
+        // has finished on the current value.
+        nextZipper = nextZipper.moveToParent();
+        finishedVisitingCurrent = true;
+        return;
+      }
+    }
+  }
+
+  /**
+   * Interface for representing a schema element as part of a traversal.
+   *
+   * <p>This object should always be treated ephemeral and not be referenced once {@code next()} is
+   * called on the iterator.
+   */
+  public interface SchemaElement {
+    /** Get the current field. */
+    StructField getField();
+
+    /** Returns the path to the node via user facing names. */
```

Comment:
Are there backticks for columns with `.`s in them? e.g. if my 1st column is `foo.bar` and my nested column is `zip.zip` will this return
```
foo.bar.zip.zip
or
`foo.bar`.`zip.zip`
```

Reply:
Ah, I see that's a TODO